core/market/src/negotiation/requestor.rs

```diff
@@ -391,6 +395,52 @@ impl RequestorBroker {
             AgreementState::Terminated => AgreementStateError::Terminated(agreement.id),
         })?
     }
+
+    pub async fn terminate_agreement(&self, id: Identity, 
+        agreement_id: &AgreementId,
+    ) -> Result<(), AgreementError> {
+        let dao = self.common.db.as_dao::<AgreementDao>();
+        let mut agreement = match dao
+            .select(
+                agreement_id,
+                Some(id.identity.clone()),
+                Utc::now().naive_utc(),
+            )
+            .await
+            .map_err(|e| AgreementError::Get(agreement_id.clone(), e))?
+        {
+            None => return Err(AgreementError::NotFound(agreement_id.clone())),
+            Some(agreement) => agreement,
+        };
+
+        Err(match agreement.state {
+            AgreementState::Proposal => AgreementStateError::Proposed(agreement.id),
+            AgreementState::Pending => AgreementStateError::Confirmed(agreement.id),
+            AgreementState::Cancelled => AgreementStateError::Cancelled(agreement.id),
+            AgreementState::Rejected => AgreementStateError::Rejected(agreement.id),
+            AgreementState::Approved => {
+                agreement.state = AgreementState::Terminated;
+                self.api.terminate_agreement(
+                    &agreement,
+                    id.identity.clone(),
+                    agreement.provider_id,
+                ).await?;
+                dao.update_state(agreement_id, AgreementState::Terminated)
+                    .await
+                    .map_err(|e| AgreementError::Get(agreement_id.clone(), e))?;
+
+                counter!("market.agreements.requestor.terminated", 1);
+                log::info!(
+                    "Requestor {} terminated Agreement [{}] and sent to Provider.",
+                    DisplayIdentity(&id),
```

Comment:
Note for future: in https://github.com/golemfactory/yagna/pull/754 you can do `id.display()`